cloud/aws/actuators/machine/actuator.go

```diff
@@ -183,6 +243,408 @@ func (a *Actuator) Exists(cluster *clusterv1.Cluster, machine *clusterv1.Machine
 	}
 }
 
+// Ensures that the security groups of the machine are correct
+// Returns bool, error
+// Bool indicates if changes were made or not, allowing the caller to decide
+// if the machine should be updated.
+func (a *Actuator) ensureSecurityGroups(machine *clusterv1.Machine, instanceID *string, additionalSecurityGroups []v1alpha1.AWSResourceReference, instanceSecurityGroups map[string]string) (bool, error) {
+	// Get the SecurityGroup annotations
+	annotation, err := a.machineAnnotationJSON(machine, SecurityGroupsLastAppliedAnnotation)
+	if err != nil {
+		return false, err
+	}
+
+	// Now we check to see if we need to mutate any mutable state.
+	// Check if any additions have been made to the instance Security Groups.
+	changed, groupIDs, newAnnotation := a.securityGroupsChanged(
+		annotation,
+		additionalSecurityGroups,
+		instanceSecurityGroups,
+	)
+	if changed {
+		// Finally, update the instance with our new security groups.
+		err := a.ec2.UpdateInstanceSecurityGroups(instanceID, groupIDs)
+		if err != nil {
+			return false, err
+		}
+
+		// And then update the annotation to reflect our new managed state.
+		err = a.updateMachineAnnotationJSON(machine, SecurityGroupsLastAppliedAnnotation, newAnnotation)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	return changed, nil
+}
+
+// Ensure that the tags of the machine are correct
+// Returns bool, error
+// Bool indicates if changes were made or not, allowing the caller to decide
+// if the machine should be updated.
+func (a *Actuator) ensureTags(machine *clusterv1.Machine, instanceID *string, additionalTags map[string]string) (bool, error) {
+	annotation, err := a.machineAnnotationJSON(machine, TagsLastAppliedAnnotation)
+	if err != nil {
+		return false, err
+	}
+
+	// Check if the instance tags were changed. If they were, update them.
+	// It would be possible here to only send new/updated tags, but for the
+	// moment we send everything, even if only a single tag was created or
+	// upated.
+	changed, created, deleted, newAnnotation := a.tagsChanged(annotation, additionalTags)
+	if changed {
+		err = a.ec2.UpdateResourceTags(instanceID, created, deleted)
+		if err != nil {
+			return false, err
+		}
+
+		// We also need to update the annotation if anything changed.
+		err = a.updateMachineAnnotationJSON(machine, TagsLastAppliedAnnotation, newAnnotation)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	return changed, nil
+}
+
+// Arguments:
```

Comment:
This comment is quite verbose and going to be hard to keep up to date. Go doc will do a good job with just a function signature and a few lines of explaining text. I think this could be reduced to something like:

```go
// securityGroupsChanged determines which security groups to delete and which to add.
```

The detail is nice but it's really hard to maintain comments like this

Reply:
Am generally in favour of over commenting vs. under commenting, at least until we've properly documented our implementation decisions and assumptions that affect consumers of cluster-api-provider-aws.